Separate quaternion algebra from particle rotation

jngrad opened the pull request:
Description of changes:
- extract quaternion algebra from the `particle_data.hpp` and `rotation.hpp` files
   - removes code duplication caused by a (now resolved) circular dependency (see #3157)
   - makes it possible to replace the quaternion code by a dedicated library, e.g. [boost:qvm::quat](https://www.boost.org/doc/libs/1_68_0/libs/qvm/doc/index.html) or [boost::math::quaternion](https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/quaternions.html) in the core (see #2289), [rowan](https://rowan.readthedocs.io/en/latest/) in the interface (see #2964)
- simplify code around call sites using Vector4d arithmetic, `std::tuple`, Particle references
- documentation cleanup

jngrad committed:
Separate quaternion algebra from particle rotation

jngrad committed:
Refactor quaternion algebra

jngrad committed:
Refactor quaternion call sites

Use Vector4d arithmetic and std::tuple, remove unused error codes,
simplify the Cython interface by enabling Vector4d indexing.

jngrad committed:
Update Doxygen and Sphinx docs

jngrad committed:
Clang-tidy

RudolfWeeber commented:
I'd suggest deferring this until after 4.1. The freeze is already a few days overdue and I'd like to focus on release-relevant issues, right now.

fweik commented on src/core/quaternion.hpp:
```diff
@@ -0,0 +1,94 @@
+
+  // Calculate angles
+  auto const d_xy = sqrt(d[0] * d[0] + d[1] * d[1]);
+  double theta2, phi2;
```
`T`

jngrad replied:
thanks!

fweik commented on src/core/quaternion.hpp:
```diff
@@ -0,0 +1,94 @@
+  auto const dm = d.norm();
+
+  // null vectors cannot be converted to quaternions
+  if (dm < ROUND_ERROR_PREC) {
```
This assumes `dm` is `double`, please use `std::numeric_limits`.

jngrad replied:
fixed

fweik commented on src/core/quaternion.hpp:
```diff
@@ -0,0 +1,94 @@
+    // We suppose that theta2 = theta/2 and phi2 = (phi - pi/2)/2,
+    // where angles theta and phi are in spherical coordinates
+    theta2 = acos(d[2] / dm) / 2;
+    phi2 = ((d[1] > 0) ? 1 : -1) * acos(d[0] / d_xy) / 2 - Utils::pi() / 4;
```
`Utils::sgn(d[1])`, `Utils::pi<T>()`

jngrad replied:
`Utils::sgn(d[1])` won't work here because it returns a value in the set {-1, 0, 1}. If `d[1]` is zero `phi2` will be -pi/2 instead of +pi/2 and the resulting quaternion will have wrong signs, causing the python test `collision_detection` to fail.

jngrad replied:
The signum edge case is now documented and tested.

fweik commented on src/core/quaternion.hpp:
```diff
@@ -0,0 +1,94 @@
+  if (d_xy == 0) {
+    // Here the director is co-linear with the z-azis
+    // We need to distinguish between (0, 0, +d_z) and (0, 0, -d_z)
+    theta2 = (d[2] > 0) ? 0 : Utils::pi() / 2;
```
`Utils::pi<T>()`

fweik commented on src/core/quaternion.hpp:
```diff
@@ -0,0 +1,94 @@
+ *  Quaternion algebra.
+ */
+
+#ifdef ROTATION
```
This is not needed, please remove. (Also observe that the header is not self-contained, because the config is not included).

jngrad replied:
fixed

fweik reviewed (changes requested):
Please move `quaternion.hpp` to utils and add unit tests for the three functions in there.

jngrad committed:
Fix template logic

jngrad commented:
@fweik I personally wouldn't move that file to utils because it's not re-usable. The corner case with +/-pi/2 shows this quaternion implementation only works because the calling sites were tailored for it. Writing a C++ unit test would not really make sense here, because only the final result of the rotation is of interest (already covered by Python tests), and it would make maintenance harder (we once had a C++ tabulated bond with the wrong sign, and a test that made sure the result had the wrong sign, because the python logic was tailored for it). Coming up with a unit test is also annoying because the quaternion formalism is not standard, and the fact that a [unit test was hard-coded inside a calling site](https://github.com/espressomd/espresso/blob/61026c799a93e10e2e0900b43b5abb785c69dbdb/src/core/virtual_sites.cpp#L119-L125) to check the rotation inversion implementation is correct every time the function is called is odd.

fweik commented:
Please add the tests.

fweik commented:
Testing the multiplication is easy, just check the multiplication table. For the conversion functions you should check that they are inverse to each other. It's sufficient that they encode the direction, not how. If anything, when there is client code that depends on a specific behavior of the function, especially it is non-standard, that gives more reason to add a test, so that behavior is kept. So you should also add a test for the sign convention.
Also the file should be moved, it's generic math and not related to MD or the espressos core, so it is a better fit for utils.

jngrad committed:
Rename function

jngrad committed:
Move quaternion algebra to utils

jngrad committed:
Add unit tests for quaternions

jngrad committed:
Remove constexpr for expression involving std::sqrt

The C++ standard specifically excludes std::sqrt from the list of
functions in cmath that are eligible for constexpr. Although some
compilers made std::sqrt a constexpr function anyway, it's not the
case of AppleClang 9.0.

KaiSzuttor commented on src/utils/tests/quaternion_test.cpp:
```diff
@@ -0,0 +1,107 @@
+          continue;
+        }
+        // here the case where j is zero is also tested
+        Vector3d input{{(double)i, (double)j, (double)k}};
```
```static_cast<double>(i)```...

jngrad replied:
thanks!

jngrad committed:
static_cast and vector operations

jngrad committed:
Merge branch 'python' into extract-quaternions

jngrad commented:
bors r=fweik
